### ISSUE TITLE
2.0 ActiveJob Adapter has a Timeout based on Kafka configuration that isn't threadsafe.

### ISSUE THREAD

**Issue by whithajess** [karafka/karafka#809]
Karafkas ActiveJob Adapter follows a synchronous process from what I can see that follows this:

1. Pull message from Kafka
2. Execute message with ActiveJob execute
3. After execute mark as consumed.

This means if the Job execution takes longer than the [`max.poll.interval.ms` - default is **300000 (5 minutes)**](https://kafka.apache.org/documentation/#consumerconfigs_max.poll.interval.ms) set on Kafka the consumer group will rebalance and assign the partition to another member.

The problem here is that ActiveJob will continue its execution there a few cases that could happen at this point:
- **Multiple Consumers** - Different consumer picks up the message off Kafka and attempts to Execute the Job at the same time.
  - Continues to rebalance and consumers keep attempting the same job repeatedly as they consistently get moved around.
- **Single Consumer** - Keeps repeating the execution of the Job despite the fact its potentially already succeeded

--------

## Expected behaviour

Some suggestions:

- ActiveJob execution is limited to a timeout below that of Kafkas `max.poll.interval.ms` configuration _(Actual not default)_
- Karafka Adapter exposes a settable timeout for ActiveJob execution
  - Documentation so end users can set it below their `max.poll.interval.ms` configuration
- Karafka Adapter marks the message as consumed before execution and deals with `failed / timedout` jobs in a different manner (potentially a `failure` topic)

## Actual behaviour

As above.

## Steps to reproduce the problem

Insert a `sleep()` longer than that of the `max.poll.interval.ms` in a _Job_.

**NB: Suggest setting `max.poll.interval.ms` to something low on Kafka to speed up testing** 

## Your setup details

```
Karafka version: 2.0.0.alpha2
Ruby version: 3.1.0
Rdkafka version: 0.11.1
Subscription groups count: 1
Workers count: 1
Application client id: default-job
Boot file: karafka.rb
Environment: development
License: LGPL-3.0
```

**Comment by mensfeld** [comment]
Yeah, I don't want to reply to all your messages with "lets deal with this in the docs and instrumentation" but I think we should :D

> Karafka Adapter marks the message as consumed before execution

This would mean we would switch from "at least once" to "probably once" and in my opinion could cause more problems because it could skip messages in case of shutdowns, OOMs and other non standard not-timeout related issues.

We could potentially wrap a job with a `Timeout#timeout` but that is expensive from a ruby VM perspective (really expensive). 

What I would suggest here (and that's actually for myself :D) is to add tracking of this into the DD instrumentation that I am building and create a dedicated dashboard from which alerts can be issued.

In general if a job is running for 5 minutes, then I would anyhow recommend a different backend than one powered by Kafka.

**Comment by pond** [comment]
> "at least once" to "probably once" 

Hmm - well, FWIW - naïvely, the way I'd approach a Kafka job adapter might be:

* Put most of the login in a job wrapper class
* When a job is received by the Kafka consumer, it launches it in a process/thread/whatever and the job wrapper code within that executing context is responsible for acknowledging the Kafka message
* This means that the Kafka message isn't acknowledged until the job has definitely started
* If the job then subsequently crashes well, so be it; we're not in the business of trying to restart crashed jobs - it crashed for a reason, restarting probably just means a crash loop! Our docs would describe exception conditions & things that the client of our job system must handle (ActiveJob already defines much of that anyway, including opt-in retry conditions)
* Since we pull just one message and acknowledge it ASAP, there's no chance of a job running more than once (assuming Kafka itself is working)
* Per-worker queue throughput is limited by the delay between reading a message, launching the worker and the worker at-startup acknowledging the message

There's the complex issue of timeouts. At the risk of over-designing it, I can think off the top of my head of three.

On-queue:

* How long should a message wait on the Kafka queue to be processed? Often, forever; but perhaps not always.
* Example: You've an hourly job launched via `cron` or similar. The consuming system is busy (or down) and the message waits for an hour. Another instance is now being scheduled. Should you cancel scheduling the new one (there's already one in the queue) or have the one in the queue expire after an hour, knowing that another instance will be launched? The former is the more "correct" because it avoids always-getting-posted-to-the-back-of-the-queue problems with new jobs, but it's the harder to implement (how does the scheduler for a new job identify that an old job of identical nature is already queued and waiting).
* This would be a feature I imagine could be deferred for later versions of whatever software I was writing - not immediately required.

Job startup:

* How long do you wait between a message being read and a message being acknowledged before determining that the job never started, try to shut down whatever thread or whatever it was meant to be executing in, and either try again or throw the message away and raise an admin alert?

Per-job:

* How long should you wait for a job to run and how to do you track that it even exists?
* Something like the job wrapper class or the job startup class could handle this. You don't start one thread/process, you start two; one is the timer for the other. If the job completes, the wrapper shuts down its own thread/process and the watchdog thread/process. If the job is stalled, the watchdog thread/process - a simple 'sleep' affair would suffice - would finish sleeping, check state & consider the watched job stalled, then take kill-on-timeout actions.

All this said, ActiveJob introduces numerous complexities that make management harder - `discard_on`, `retry_on` etc. are potentially troublesome depending on how much support in the Rails base class exists & how exactly it's all meant to tie together.

But either way, however I decided to do this, the primary concerns would be that (1) the message which causes a job to start _does not_ stay on the queue while the job runs because that's not good for Kafka throughput and exposes multiple ways in which the job might accidentally get re-run, even if the scheduler is saying it shouldn't be retried; and (2) the message isn't acknowledged off Kafka until the job is actually executing, because the ack happens within the context of the executing job itself.

> We could potentially wrap a job with a `Timeout#timeout` but that is expensive

Yes, I'd not do that; Timeout is troublesome for a lot of reasons beyond expense, too. It's usually considered harmful except for carefully curated special case code and using it on general purpose externally-written job code would cause problems. Instead, as I mentioned, I think I'd go with a watchdog timer that could be (almost) as simple as a sleep followed by kill-job-on-wake. There would be some inter-process management between it and the watched job to consider. I wouldn't want things too heavyweight, but I'd be loathe to use a Ruby Thread for that because there are then ways in which the job could stall which would also prevent Thread context switches by Ruby itself, so the watchdog would never trigger; I think a watcher process would be the cleaner way.

**Comment by whithajess** [comment]
Im definitely going to have to disagree on the documentation only approach for this one. 

One of the reasons ruby `Timeout` is considered so troublesome - is the exact reason I am raising the issue.

> The main problem with using Timeout in network-related code is that timeout exceptions aren't raised on the other end of the connection, which may in turn send messages back to your application to be read in the wrong context. This causes unforeseen and hard to debug errors in your application. - ["Timeout is Ruby's Most Dangerous API" —Mike Perham](https://www.mikeperham.com/2015/05/08/timeout-rubys-most-dangerous-api/)

While I am a massive fan of DD (I originally wrote the cross application tracing for the dd-trace-rb gem back in 0.11.0) this will always be one of those issues that will be an absolute nightmare to debug when it occurs.

Remember _5 minutes_ isn't a guaranteed number in high performing Kafka clusters I have seen this set much lower so things like instant messages don't get held up by a dead consumer - its a Kafka wide config

Aside from the above mention of longer running jobs from @pond quite even if you never intend to go over **5 (maybe)** minutes when you inevitably do go over this mark all of a sudden you have all the `consumers` in that `consumer group` repeatedly being marked as `dead` getting rebalanced processing from a different `partition` in the meantime then getting rebalanced back onto the `partition` holding the job thats greater than a `5m (maybe)` run and getting marked as `dead` and repeating the process. Now I have `offset lag` on that `topic` but it is still processing messages and the consumer group is consuming from that topic and most of the time it has the correct amount of consumers as well - Making it almost impossible to find the bug thats only ever really referenced in some documentation on the Karafka gem which is one of many consumer types in the overall system.

**Comment by whithajess** [comment]
I originally tested the Consumer and didn't notice the less obvious than ActiveJob bug around the same issue^

**Comment by mensfeld** [comment]
Redirecting this to 2.1. There will be a solution as async (non blocking) jobs are already implemented.

**Comment by mensfeld** [comment]
closing in favour of LRJ for both AJ and non-AJ: https://github.com/karafka/karafka/issues/866

**Comment by mensfeld** [comment]
feel free to give it a shot: https://github.com/karafka/karafka/issues/866
